### ISSUE TITLE
Building libzstd.a using NDK is broken

### ISSUE THREAD

**Issue by YCJDWCJI** [facebook/zstd#4103]
**Describe the bug:**

Since commit [345bcb5], building with NDK is broken, error:

```Text
dictBuilder/cover.c:332:5: error: call to undeclared function 'qsort_r'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
```

[345bcb5]: https://github.com/facebook/zstd/blob/345bcb5ff7001fbe2dfd59974808170c9e0f4d5d/lib/dictBuilder/cover.c

**To reproduce:**

1. Configure [NDK toolchain].
2. `git clone https://github.com/facebook/zstd`.
3. `cd zstd/lib`. 
4. `make libzstd.a`.
5. See error.

[NDK toolchain]: https://developer.android.com/ndk/guides/other_build_systems

**Expected behavior:**

The build process should complete successfully and produce the **libzstd.a** file.

**Desktop:**

 - OS: [Arch Linux] [Testing]
 - Compiler: aarch64-linux-android31-clang
 - Build system: makefile

[Arch Linux]: https://archlinux.org
[Testing]: https://wiki.archlinux.org/title/Official_repositories#Testing_repositories

**Temporary Solution:**

Using **C90 fallback** under the `_GNU_SOURCE` macro.

**Comment by Cyan4973** [comment]
cc @Adenilson

**Comment by Adenilson** [comment]
I will proceed to repro and fix this one.

**Comment by Adenilson** [comment]
I managed to repro the bug report by installing the NDK r27 (https://developer.android.com/ndk/downloads) and trying to build by exporting CC and CXX variables to point to the aarch64 cross compiler.

The output was:

//-------------------------------------------------------------------------------------------------------------------------------------------
CC obj/conf_e35c1c02789d063e7f14d8327851243f/dynamic/divsufsort.o
/home/adenilson/zstd_fork/lib//dictBuilder/cover.c:332:5: error: call to undeclared function 'qsort_r'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  332 |     qsort_r(ctx->suffix, ctx->suffixSize, sizeof(U32),
      |     ^
/home/adenilson/zstd_fork/lib//dictBuilder/cover.c:332:5: note: did you mean 'qsort'?
/home/adenilson/android-ndk-r27/toolchains/llvm/prebuilt/linux-x86_64/bin/../sysroot/usr/include/stdlib.h:126:6: note: 'qsort' declared here
  126 | void qsort(void* _Nullable __base, size_t __nmemb, size_t __size, int (* _Nonnull __comparator)(const void* _Nullable __lhs, const void* _Nullable __rhs));
      |      ^
1 error generated.
//-------------------------------------------------------------------------------------------------------------------------------------------


Now I will proceed to identify what is the underlying issue and write a fix for it.

**Comment by Adenilson** [comment]
From the looks of it, even though the NDK uses llvm, it is missing qsort_r().

The compiler defines __linux which is triggering the usage of qsort_r() by our code.

We can test for the additional __ANDROID__ macro and disable it accordingly.

Unrelated to zstd, but it is a bit puzzling why the NDK doesn't include qsort_r(), given it is available on both gcc/glibc and llvm/libc.

I will dig a bit more and see what I can find out.

**Comment by Adenilson** [comment]
Something like this seems to do the trick:
https://gist.github.com/Adenilson/1133c8422d43d7b55e9565805b8fdff1

I will do further testing and validation and should next upload a fix.

**Comment by Cyan4973** [comment]
I was wondering if this issue would be ground for an additional test in CI

**Comment by Adenilson** [comment]
I got admit that I was caught by surprise on this one. I mean, clang/libc has what is needed to provide qsort_r(), but for some reason that is not exported through the NDK.

I will make sure that any future patches I contribute will be validated with the NDK.
:-)

Given that Android/NDK is a supported platform by zstd, I feel it would make sense in the future if an Android/NDK bot was part of the infrastructure to catch this kind of bugs early on.

**Comment by Cyan4973** [comment]
issue fixed in #4107
